src/boostvqe/models/dbi/double_bracket_evolution_oracles.py

```diff
@@ -307,3 +307,43 @@ def circuit(self, t_duration, steps=None, order=None):
             steps=steps,
             order=order,
         )
+
+
+@dataclass
+class tfim_EvolutionOracle(EvolutionOracle):
+    steps: int = None
+    B_a: float = None
+
+    def circuit(self, a, t_duration, B_a, steps=None, order=None):
+        if steps is None:
+            steps = self.steps
+
+        circuit = Circuit(self.h.nqubits)  # Initialize the circuit with the number of qubits
+
+        # Add CNOT(a, a+1)
+        circuit.add(gates.CNOT(a, a + 1))
+
+        # Time evolution under the transverse field Ising model Hamiltonian
+        # exp(-i t (X(a) + B_a * Z(a)))
+        dt = t_duration / steps  # Divide the time duration for Trotterization if needed
```

Comment:
```suggestion
        dt = t_duration / steps  # Divide the time duration for Trotterization if needed
        
        for a in range(nqubits:
        
        circuit.add(gates.CNOT(a, a + 1))
        circuit += self._time_evolution_step(a, dt, B_a)
        circuit.add(gates.CNOT(a, a + 1))
```

and then loop this over _ in range(steps)

Reply:
'
n_qubits = 3
h_coeff = 1
hamiltonian = SymbolicHamiltonian(nqubits=n_qubits)

oracle = TFIM_EvolutionOracle(h=hamiltonian, evolution_oracle_type="trotter", steps=1, B_a=0.8, order=1)

circuit = oracle.circuit(t_duration=1.0)

oracle = TFIM_EvolutionOracle(h=hamiltonian, evolution_oracle_type="trotter", steps=1, B_a=0, order=2)

circuit = oracle.circuit(t_duration=1.0)

unitary = circuit.unitary()


from qibo import hamiltonians
from numpy.linalg import norm

def our_TFIM(nqubits, h: float = 0.0, dense: bool = True, backend=None):
    def multikron(matrix_list):
        """Calculates Kronecker product of a list of matrices."""
        return reduce(np.kron, matrix_list)

    from qibo.backends import matrices

    matrix = (
        - multikron([matrices.X, matrices.X]) - h * multikron([matrices.Z, matrices.I])
    )
    terms = [hamiltonians.terms.HamiltonianTerm(matrix, i, i + 1) for i in range(nqubits - 1)]
    terms.append(hamiltonians.terms.HamiltonianTerm(matrix, nqubits - 1, 0))
    ham = SymbolicHamiltonian(backend=backend)
    ham.terms = terms
    return ham

ham = our_TFIM(nqubits=n_qubits, h=h_coeff, dense=False)
truth = ham.exp(1)
verification_norm = []
for step in range(1, 21):
    oracle = TFIM_EvolutionOracle(h=hamiltonian, evolution_oracle_type="trotter", steps=step, B_a=h_coeff, order=2)
    circuit = oracle.circuit(t_duration=1.0)
    unitary = circuit.unitary()
    #print(norm(truth-unitary))
    verification_norm.append(norm(truth-unitary))

import matplotlib.pyplot as plt
x = np.array([i for i in range(1, 21)])
plt.plot(x, verification_norm, 'o')
plt.title("verification of TFIM 2 CNOT implementation")
plt.xlabel("steps")
plt.ylabel("norm of difference")

plt.show()`

produces the following graph:

<img width="626" alt="Screenshot 2024-11-20 at 3 19 27 PM" src="https://github.com/user-attachments/assets/b9158b63-da69-454e-a87b-af7c7c1805e3">